File: 2021/02/navigate.py
def read_input():
    with open('../input/02.txt') as f:
        input = f.read().split()
        return [(input[i], int(input[i+1])) for i in range(0, len(input), 2)]

def navigate(input):
    position, depth = 0, 0
    for cmd, v in input:
        if cmd == 'forward':
            position += v
        elif cmd == 'down':
            depth += v
        elif cmd == 'up':
            depth -= v
    return position * depth

def navigate2(input):
    position, depth, aim = 0, 0, 0
    for cmd, v in input:
        if cmd == 'forward':
            position += v
            depth += v * aim
        elif cmd == 'down':
            aim += v
        elif cmd == 'up':
            aim -= v
    return position * depth

def main():
    input = read_input()
    print(navigate(input))
    print(navigate2(input))

if __name__ == "__main__":
    main()

File: 2021/02/navigate.rb
def navigate(input)
  reduced = input.each_slice(2).each_with_object(Hash.new(0)) { |(k, v), m| m[k] += v }

  reduced['forward'] * (reduced['down'] - reduced['up'])
end

def navigate2(input)
  position, depth, aim = 0, 0, 0

  input.each_slice(2) do |(cmd, v)|
    case cmd
    when 'forward'
      position += v
      depth += v * aim
    when 'down'
      aim += v
    when 'up'
      aim -= v
    end
  end

  position * depth
end

input = File.read('../input/02.txt').split().map.with_index { |e, i| i.odd? ? e.to_i : e }

p navigate(input)
p navigate2(input)
